src/query/api/v1/handler/prometheus/native/list_tags_test.go

```diff
@@ -206,3 +208,40 @@ func TestListErrorTags(t *testing.T) {
 		require.JSONEq(t, `{"status":"error","error":"err"}`, string(r))
 	}
 }
+
+//nolint:dupl
+func TestListTagsTimeout(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	// Setup storage
+	store := storage.NewMockStorage(ctrl)
+	store.EXPECT().CompleteTags(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(ctx context.Context,
+			query *storage.CompleteTagsQuery,
+			options *storage.FetchOptions,
+		) (*consolidators.CompleteTagsResult, error) {
+			<-ctx.Done()
+			// Prove the fact that we've passed along the timeout and that it's exceeded
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				return nil, ctx.Err()
+			}
+			return nil, nil
+		},
+	)
+
+	fb, err := handleroptions.NewFetchOptionsBuilder(
+		handleroptions.FetchOptionsBuilderOptions{Timeout: 1 * time.Millisecond})
+	require.NoError(t, err)
+	opts := options.EmptyHandlerOptions().
+		SetStorage(store).
+		SetFetchOptionsBuilder(fb)
+
+	req := httptest.NewRequest("GET", "/labels", nil)
+	w := httptest.NewRecorder()
+	h := NewListTagsHandler(opts)
+	h.ServeHTTP(w, req)
+
+	assert.Equal(t, 504, w.Code, "Status code not 504")
```

Comment:
Two questions:
- Why a 504 here and a 500 previously?
- A timeout should most likely be a 4xx ya -- (maybe a 408)?

Realize we may not want to resolve in this PR but just asking the question

Reply:
timeouts are definitely a grey area between 4xx and 5xx. did the client issue a long query (4xx-ish)? are the servers slow resulting in a normal query taking longer (5xx-ish)? 

I don't think 408 is appropriate for this case.

Reply:
Yeah, that's a good point. We probably want to err on the side of 500 given the current state of the server :P